### PR TITLE
Feat/bytecode interpreter tests

### PR DESCRIPTION
**Pull request by guillep**
Cleanups in context and instruction stream
 - cut the inheritance between InstructionStream and context
 -  review the API used by the debugger
 - make Context be a normal client of the instruction stream
 - remove unused methods
 - comment methods

### CODE REVIEW

**Comment by Ducasse** [comment]
Guille did you edit some files by hand? Because the build break in the 

```
STON class>>fromStream:
STON class>>fromString:
TonelParser>>metadata
[ self metadata ] in TonelParser>>typeDef in Block: [ self metadata ]
[ ^ successBlock value: aBlock value ] in TonelParser>>try:onSuccess:onFailure: in Block: [ ^ successBlock value: aBlock value ]
BlockClosure>>on:do:
TonelParser>>try:onSuccess:onFailure:
TonelParser>>try:
TonelParser>>typeDef
TonelParser>>document
TonelParser>>start
TonelParser class>>parseStream:
[ :s | TonelParser parseStream: s ] in [ :each | each readStreamDo: [ :s | TonelParser parseStream: s ] ] in TonelReader>>loadDefinitions in Block: [ :s | TonelParser parseStream: s ]
[ aBlock value: stream ] in FileReference(AbstractFileReference)>>readStreamDo: in Block: [ aBlock value: stream ]
BlockClosure>>ensure:
FileReference(AbstractFileReference)>>readStreamDo:
2021-07-09T18:04:17.296785+02:00 PBStepStartedSignal Starting  Process.
```

**Comment by Ducasse** [comment]
I read all the code of the diff but I could not spot something trivial.

**Comment by Ducasse** [comment]
```
Errors in script loaded from /builds/workspace/uest_and_branch_Pipeline_PR-9573/64/bootstrap/scripts/bootstrap.st'
STONReaderError: At character 77: , expected
STONReader>>error:
STONReader>>expectChar:
```

**Comment by Ducasse** [comment]
```
TON class>>fromString:
	Receiver: STON
	Arguments and temporary variables: 
		string: 	'{
	#name : #Context,
	#type : #variable,
	#instVars : [
		''sender'',
...etc...
	Receiver's instance variables: 
		superclass: 	Object
		methodDict: 	a MethodDictionary()
		format: 	0
		layout: 	a FixedLayout
		organization: 	a ClassOrganization
		subclasses: 	nil
		name: 	#STON
		classPool: 	a Dictionary()
		sharedPools: 	an OrderedCollection()
		environment: 	a SystemDictionary(lots of globals)
		category: 	#'STON-Core-Base'
```

**Comment by Ducasse** [comment]
I read all https://github.com/pharo-project/pharo/blob/601d659a69e502fa8b5ac55430df3a9c26da4609/src/Kernel/Context.class.st but nothing jump to my eye. The class definition looks correct.


<blockquote><div> GitHub</div><div><strong><a href="https://github.com/pharo-project/pharo">pharo-project/pharo</a></strong></div><div>Pharo is a dynamic reflective pure object-oriented language supporting live programming inspired by Smalltalk. - pharo-project/pharo</div></blockquote>

**Comment by Ducasse** [comment]
I found it :)

```
Class {
	#name : #Context,
	#type : #variable,
	#instVars : [
		'sender',
		'pc'
		'stackp',
```

**Comment by guillep** [comment]
We should open an issue to make parse errors print some more context ;)

**Comment by Ducasse** [comment]
Yes I thought the same.

**Comment by Ducasse** [comment]
```

KeyNotFound: key #superclass not found in Dictionary
```
I will add it.

**Comment by guillep** [comment]
There is another error:

```smalltalk
UndefinedObject>>handleSignal:
MessageNotUnderstood(Exception)>>signal
UndefinedObject(Object)>>doesNotUnderstand: #check:
UndefinedObject>>doesNotUnderstand: #check:
BlockClosure>>isClean
```

Which seems to be a problem in the order of dependencies. It seems that the class CleanBlockChecker is not loaded in the place where `isClean` is used.

```smalltalk
isClean
	"Answer if the receiver does not close-over any variables other than globals, and does
	 not ^-return (does not close over the home context).  Clean blocks are amenable to
	 being created at compile-time."
	self numCopiedValues > 0 ifTrue:
		[^false].

	^ CleanBlockChecker check: self method
```

File: src/Kernel/Object.class.st [hunk 1]
```diff
@@ -496,6 +496,12 @@ Object >> bindingOf: aString [
 	^nil
 ]
 
+{ #category : #comparing }
+Object >> bla [
+
+	^ 1
+]
+
```
Comment (guillep) [inline src/Kernel/Object.class.st]: This is a debugging leftover, we should remove it.

File: src/Kernel/Object.class.st [hunk 2]
```diff
@@ -1993,6 +1999,12 @@ Object >> toFinalizeSend: aSelector to: aFinalizer with: aResourceHandle [
 			argument: aResourceHandle)
 ]
 
+{ #category : #comparing }
+Object >> toto [
+
+	self bla
+]
+
```
Comment (guillep) [inline src/Kernel/Object.class.st]: same here

**Comment by Ducasse** [comment]
Now that clean blocks are part of the kernel let us see.

**Comment by Ducasse** [comment]
We get another problem

```
Context>> #pc:
Context(Object)>>doesNotUnderstand: #pc:
Process>>doTerminationFromAnotherProcess
Process>>terminate
SmalltalkImage>>installLowSpaceWatcher
ProcessorScheduler class>>startUp
ProcessorScheduler class(Behavior)>>startUp:
ClassSessionHandler>>startup:
<Error printing blockClosure in: WorkingSession>>#runStartup: > in WorkingSession>>runStartup: in Block: <Error printing blockClosure in: WorkingSession>>#...etc...
<Error printing blockClosure in: WorkingSession>>#runList:do: > in <Error printing blockClosure in: WorkingSession>>#runList:do: > in WorkingSession>>runList:do: in Block: <Error printing blockClosure in: WorkingSession>>#...etc...
BlockClosure>>on:do:
<Error printing blockClosure in: WorkingSession>>#runList:do: > in WorkingSession>>runList:do: in Block: <Error printing blockClosure in: WorkingSession>>#...etc...
Array(SequenceableCollection)>>do:
WorkingSession>>runList:do:
WorkingSession>>runStartup:
WorkingSession>>start:
SessionManager>>launchSnapshot:andQuit:
<Error printing blockClosure in: SessionManager>>#snapshot:andQuit: > in SessionManager>>snapshot:andQuit: in Block: <Error printing blockClosure in: SessionManager>>#...etc...
<Error printing blockClosure in: BlockClosure>>#newProcess > in BlockClosure>>newProcess in Block: <Error printing blockClosure in: BlockClosure>>#ne...etc...
SystemNotification: [Hermes] Reading Hermes-Extensions.hermes
SystemNotification: [Hermes] Installing Hermes-Extensions.hermes
SystemNotification: [Hermes] Remaining Undeclared variables in the system: #(#FileLocator #FileAttributeNotSupported #Unicode #StartupPreferencesLoader #BlockLocalTempCounter #FileSystem)

Process>>terminate
SmalltalkImage>>installLowSpaceWatcher
ProcessorScheduler class>>startUp
ProcessorScheduler class(Behavior)>>startUp:
ClassSessionHandler>>startup:
<Error printing blockClosure in: WorkingSession>>#runStartup: > in WorkingSession>>runStartup: in Block: <Error printing blockClosure in: WorkingSession>>#...etc...
<Error printing blockClosure in: WorkingSession>>#runList:do: > in <Error printing blockClosure in: WorkingSession>>#runList:do: > in WorkingSession>>runList:do: in Block: <Error printing blockClosure in: WorkingSession>>#...etc...
BlockClosure>>on:do:
<Error printing blockClosure in: WorkingSession>>#runList:do: > in WorkingSession>>runList:do: in Block: <Error printing blockClosure in: WorkingSession>>#...etc...
Array(SequenceableCollection)>>do:
WorkingSession>>runList:do:
WorkingSession>>runStartup:
WorkingSession>>start:
SessionManager>>launchSnapshot:andQuit:
<Error printing blockClosure in: SessionManager>>#snapshot:andQuit: > in SessionManager>>snapshot:andQuit: in Block: <Error printing blockClosure in: SessionManager>>#...etc...
<Error printing blockClosure in: BlockClosure>>#newProcess > in BlockClosure>>newProcess in Block: <Error printing blockClosure in: BlockClosure>>#ne...etc...
+ /builds/workspace/uest_and_branch_Pipeline_PR-9573/64/bootstrap-cache/vmtarget/pharo --headless Pharo9.0-PR-compiler-64bit-962a4da.image --no-default-preferences loadHermes Math-Operations-Extensions.hermes Debugging-Core.hermes Kernel-Chronology-Extras.hermes Multilingual-Encodings.hermes ReflectionMirrors-Primitives.hermes --save --no-fail-on-undeclared
pthread_setschedparam failed: Operation not permitted
This VM uses a separate heartbeat thread to update its internal clock
and handle events.  For best operation, this thread should run at a
higher priority, however the VM was unable to change the priority.  The
effect is that heavily loaded systems may experience some latency
issues.  If this occurs, please create the appropriate configuration
file in /etc/security/limits.d/ as shown below:

cat <<END | sudo tee /etc/security/limits.d/pharo.conf
*      hard    rtprio  2
*      soft    rtprio  2
END

and report to the pharo mailing list whether this improves behaviour.

```

**Comment by Ducasse** [comment]
```
SystemNotification: Loading Kernel-tonel.1.mcz
Instance of Context did not understand #pc
Context(Object)>>doesNotUnderstand: #pc
Process>>isTerminated
DelaySemaphoreScheduler(DelayBasicScheduler)>>stopTimerEventLoop
Delay class>>initialize
MCMethodDefinition>>postloadOver:
[ :each | each postloadOver: (self obsoletionFor: each) ] in MCPackageLoader>>basicLoadDefinitions in Block: [ :each | each postloadOver: (self obsoletionFor: ...etc...
```

**Comment by Ducasse** [comment]
Next error 

```
Instance of Context did not understand #willSend
```

**Comment by Ducasse** [comment]
@guillep we broke 283 tests of the decompiler so it should be something simple because it breaks all the tests :)

**Pull request merged**